pkg/input/joystick_controller.go

```diff
@@ -22,24 +24,52 @@ type JoystickController struct {
 	mapping         *GameControllerMapping
 }
 
+var (
+	JoystickControllers map[int]*JoystickController
+)
+
+func init() {
+	// Map keeping track of the connected joysticks
+	JoystickControllers = make(map[int]*JoystickController, MAX_NUM_JOYSTICKS)
+
+	// Attach the status change callback
+	glfw.SetJoystickCallback(func(joy, event int) {
+		if glfw.MonitorEvent(event) == glfw.Connected {
+			// The joystick was connected
+			log.Printf("Joystick #%d: plugged in", joy)
+			if JoystickControllers[joy] != nil {
+				JoystickControllers[joy].Open(joy)
+			}
+		} else if glfw.MonitorEvent(event) == glfw.Disconnected {
+			// The joystick was disconnected
+			log.Printf("Joystick #%d: plugged out", joy)
+			if JoystickControllers[joy] != nil && JoystickControllers[joy].Connected() {
+				JoystickControllers[joy].pluggedOut()
+			}
+		}
+	})
+}
+
 func (c *JoystickController) Open(deviceIndex int) bool {
 	if c.connected {
-		if glfw.Joystick(deviceIndex) == c.joystick {
-			// Device already connected
-			return true
-		} else {
-			// We are opening another device
-			c.Close()
-		}
+		log.Printf("Joystick already open on device #%d", c.joystick)
+		return true
 	}
 
-	if !glfw.JoystickPresent(glfw.Joystick(deviceIndex)) {
-		c.Close()
+	if JoystickControllers[deviceIndex] != nil && JoystickControllers[deviceIndex] != c {
+		log.Printf("Another joystick is associated to index #%d", deviceIndex)
 		return false
 	}
 
-	c.connected = true
+	JoystickControllers[deviceIndex] = c
 	c.joystick = glfw.Joystick(deviceIndex)
+
+	// The joystick is currently not connected but it might be plugged in later
+	if !glfw.JoystickPresent(glfw.Joystick(deviceIndex)) {
```

Comment:
This is one of the biggest changes. Even if the joystick is not present, this method returns true. It's going to connect as soon as a physical device is plugged in